dali/kernels/imgproc/color_manipulation/debayer/debayer_test.cc

```diff
@@ -0,0 +1,178 @@
+  using Kernel = NppDebayerKernel<InOutT>;
+  static constexpr int num_channels = 3;
+  static_assert(DebayerTestParamsT::alg == DALIDebayerAlgorithm::DALI_DEBAYER_BILINEAR_NPP);
+
+  void FillWithGradient(TensorListView<StorageCPU, InOutT, 3> rgb_batch) {
+    int max_val = std::numeric_limits<InOutT>::max();
+    static constexpr int num_channels = 3;
+    const auto &batch_shape = rgb_batch.shape;
+    for (int sample_idx = 0; sample_idx < rgb_batch.num_samples(); sample_idx++) {
+      int height = batch_shape[sample_idx][0];
+      int width = batch_shape[sample_idx][1];
+      auto rgb_sample = rgb_batch[sample_idx];
+      for (int h = 0; h < height; h++) {
+        for (int w = 0; w < width; w++) {
+          rgb_sample.data[h * num_channels * width + w * num_channels] = max_val * (w + 1) / width;
+        }
+      }
+      for (int h = 0; h < height; h++) {
+        for (int w = 0; w < width; w++) {
+          rgb_sample.data[h * num_channels * width + w * num_channels + 1] =
+              max_val * (h + 1) / height;
+        }
+      }
+      for (int h = 0; h < height; h++) {
+        for (int w = 0; w < width; w++) {
+          rgb_sample.data[h * num_channels * width + w * num_channels + 2] =
+              max_val * (width - w) / width;
+        }
+      }
+    }
+  }
+
+  void BayerSamples(TensorListView<StorageCPU, InOutT, 2> bayer_batch,
+                    TensorListView<StorageCPU, const InOutT, 3> rgb_batch) {
+    // Note that dali uses opncv's convention of naming the patterns, which
+    // looks at the 2x2 tile starting at the second row and column of the sensors matrix.
+    // We iterate over whole image, so the patterns are first transformed as if by looking
+    // at the tile starting at first column and row.
+    std::array<std::array<std::array<int, 2>, 2>, 4> pattern2channel{{
+        {{{0, 1}, {1, 2}}},  // bggr -> rggb -> 0112
+        {{{1, 0}, {2, 1}}},  // gbrg -> grbg -> 1021
+        {{{1, 2}, {0, 1}}},  // grbg -> gbrg -> 1201
+        {{{2, 1}, {1, 0}}}   // rggb -> bggr -> 2110
+    }};
+    int batch_size = rgb_batch.num_samples();
+    const auto &batch_shape = rgb_batch.shape;
+    for (int sample_idx = 0; sample_idx < batch_size; sample_idx++) {
+      int height = batch_shape[sample_idx][0];
+      int width = batch_shape[sample_idx][1];
+      auto pattern = patterns_[sample_idx];
+      auto rgb_sample = rgb_batch[sample_idx];
+      auto bayer_sample = bayer_batch[sample_idx];
+      ASSERT_EQ(height % 2, 0);
+      ASSERT_EQ(width % 2, 0);
+      for (int h = 0; h < height; h += 2) {
+        for (int w = 0; w < width; w += 2) {
+          for (int i = 0; i < 2; i++) {
+            for (int j = 0; j < 2; j++) {
+              bayer_sample.data[(h + i) * width + w + j] =
+                  rgb_sample.data[(h + i) * width * num_channels + (w + j) * num_channels +
+                                  pattern2channel[static_cast<int>(pattern)][i][j]];
+            }
+          }
```

Comment:
Perhaps this is simpler? Your call.
```suggestion
          int i = y & 1;
          int j = x & 1;
          int c = pattern2channel[static_cast<int>(pattern)][i][j];
          bayer_sample.data[h * width + w] =
              rgb_sample.data[(h * width + w) * num_channels) * num_channels + c];
```
BTW, isn't num_channels always 3?

Reply:
Fair enough, it is simpler. As to the number of channels, it is. I made it `static constexpr int num_channels = 3;` to "track usages".